dali/image/tiff_libtiff_impl.cc

```diff
@@ -0,0 +1,308 @@
+  } while (0)
+
+namespace dali {
+
+namespace detail {
+
+// Extracted and adjusted from OpenCV's modules/imgcodecs/src/grfmt_tiff.cpp
+class BufDecoderHelper {
+ private:
+  span<const uint8_t> &buf_;
+  size_t &buf_pos_;
+
+ public:
+  BufDecoderHelper(span<const uint8_t> &buf, size_t &buf_pos)
+      : buf_(buf), buf_pos_(buf_pos) {}
+
+  static tmsize_t read(thandle_t handle, void *buffer, tmsize_t n) {
+    BufDecoderHelper *helper = reinterpret_cast<BufDecoderHelper *>(handle);
+    auto &buf = helper->buf_;
+    const tmsize_t size = buf.size();
+    tmsize_t pos = helper->buf_pos_;
+    if (n > (size - pos)) {
+      n = size - pos;
+    }
+    memcpy(buffer, buf.data() + pos, n);
+    helper->buf_pos_ += n;
+    return n;
+  }
+
+  static tmsize_t write(thandle_t /*handle*/, void * /*buffer*/, tmsize_t /*n*/) {
+    // Not used for decoding.
+    return 0;
+  }
+
+  static toff_t seek(thandle_t handle, toff_t offset, int whence) {
+    BufDecoderHelper *helper = reinterpret_cast<BufDecoderHelper *>(handle);
+    auto &buf = helper->buf_;
+    const toff_t size = buf.size();
+    toff_t new_pos = helper->buf_pos_;
+    switch (whence) {
+    case SEEK_SET:
+      new_pos = offset;
+      break;
+    case SEEK_CUR:
+      new_pos += offset;
+      break;
+    case SEEK_END:
+      new_pos = size + offset;
+      break;
+    }
+    new_pos = std::min(new_pos, size);
+    helper->buf_pos_ = static_cast<size_t>(new_pos);
+    return new_pos;
+  }
+
+  static int map(thandle_t handle, void **base, toff_t *size) {
+    BufDecoderHelper *helper = reinterpret_cast<BufDecoderHelper *>(handle);
+    auto &buf = helper->buf_;
+    *base = const_cast<uint8_t*>(buf.data());
+    *size = buf.size();
+    return 0;
+  }
+
+  static toff_t size(thandle_t handle) {
+    BufDecoderHelper *helper = reinterpret_cast<BufDecoderHelper *>(handle);
+    return helper->buf_.size();
+  }
+
+  static int close(thandle_t handle) {
+    BufDecoderHelper *helper = reinterpret_cast<BufDecoderHelper *>(handle);
+    delete helper;
+    return 0;
+  }
+};
+
+}  // namespace detail
+
+TiffImage_LibtiffImpl::TiffImage_LibtiffImpl(const uint8_t *encoded_buffer, size_t length, DALIImageType image_type)
+    : GenericImage(encoded_buffer, length, image_type)
+    , buf_({encoded_buffer, static_cast<ptrdiff_t>(length)})
+    , buf_pos_(0) {
+  tif_.reset(
+    TIFFClientOpen("", "r",
+                   reinterpret_cast<thandle_t>(
+                     new detail::BufDecoderHelper(buf_, buf_pos_)),
+                   &detail::BufDecoderHelper::read,
+                   &detail::BufDecoderHelper::write,
+                   &detail::BufDecoderHelper::seek,
+                   &detail::BufDecoderHelper::close,
+                   &detail::BufDecoderHelper::size,
+                   &detail::BufDecoderHelper::map,
+                   /*unmap=*/0));
+
+  LIBTIFF_CALL(
+    TIFFGetField(tif_.get(), TIFFTAG_IMAGELENGTH, &shape_[0]));
```

Comment:
Shouldn't those be set in PeekDims?

Reply:
It makes life easier to do it here since I don't need to keep track of whether I've read the dimensions already (in PeekDims, DecodeImpl, etc).